1-Draft/RFCNNNN-Automatic-Resource-Management.md

```diff
@@ -0,0 +1,197 @@
+---
+RFC: RFC<four digit unique incrementing number assigned by Committee, this shall be left blank by the author>
+Author: Ilya Sazonov
+Status: Draft
+SupercededBy: <link to another RFC>
+Version: 1.0
+Area: Engine
+Comments Due: 7/1/2020
+Plan to implement: Maybe
+---
+
+# Automatic resource management
+
+## Rationale
+
+Currently PowerShell has limited capabilities compared with C# to excecute a mandatory code.
+
+PowerShell users can benefit from `finally` block of `try - catch` statement. It is very limited experience because resources in PowerShell can be distributed across scopes, cmdlets, pipelines and runspaces.
+
+If an user created a temporary file one should worry about deleting it, while PowerShell could register a mandatory code to delete this file and execute it automatically at the right time.
+
+PowerShell makes many smart things for users and could offer smart resource management.
+
+## Motivation
+
+As a script writer, I can use some well-known API-s which grabs resources
+and I want PowerShell understands this and releases the resources automatically at the right time
+so that avoids resorce leaks.
+
+As an advanced module and script writer, I can use some API-s which grabs resources
+and I want to register a code PowerShell must run to release the resources automatically at the right time so that avoids resorce leaks.
+
+As an advanced module and script writer, I want to make sure that a mandatory (cleanup, logging) code is always executed even if a pipeline / script / cmdlet is unexpectedly interrupted.
+
+As binary module developer I want to benefit from this in binary modules too.
+
+## User Experience
+
+User has no needs explicitly dispose an object when out of the script block:
```

Comment:
This basically sounds like RAII or the way Rust's borrow checker works, except without any regard for referential analysis or heap lifetimes. I personally don't think PowerShell is the right place to graft such a concept on.